platform/pulpcore/app/models/storage.py

```diff
@@ -133,12 +156,42 @@ def unlink(path):
 
         Args:
             path (str): Path to unlink.
+
+        Raises:
+            OSError: On error.
         """
         try:
             os.unlink(path)
-        except OSError as e:
-            if e.errno != errno.ENOENT:
-                raise
+        except FileNotFoundError:
+            pass
+
+    @staticmethod
+    def copy(path, content):
+        """
+        Atomic copy of the content into storage.
+        The content copied into a temporary file, then renamed.
+        This prevents creating partial files.
+
+        Args:
+            path(str): The storage path.
+            content (File): Source file object.
+
+        Raises:
+            FileExistsError: File already stored.
+            OSError: On error.
+        """
+        tmp_path = '.'.join((path, str(uuid4()), 'tmp'))
+        try:
+            with open(tmp_path, 'wb+') as fp:
+                while True:
+                    bfr = content.read(1024000)
+                    if bfr:
+                        fp.write(bfr)
+                    else:
+                        break
+            os.link(tmp_path, path)
```

Comment:
WIll this raise `errno.EXDEV`? This only runs if we are on a remote system, right? I'm guessing we want `os.rename`.

Reply:
A comment could indicate that this is the moment that the race condition is resolved.

Reply:
No, EXDEV will never be raised here.  The tmp_path and path will always be on the same filesystem.  

As it turns out, `os.rename()` has some nasty behavior in that if the target already exists, it replaces it instead of raising `FileExistsError` so it cannot be used.

Reply:
Using `shutil.move()` trades 2 lines of code for 2 lines of code and adds a dependency.  Just using the standard lib for this seems better.